// @gorhom/animated-tabbar/lib/commonjs/presets/bubble/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_ITEM_CONTAINER_WIDTH = exports.DEFAULT_ITEM_LAYOUT_DIRECTION = exports.DEFAULT_ITEM_ICON_SIZE = exports.DEFAULT_ITEM_OUTER_SPACE = exports.DEFAULT_ITEM_INNER_SPACE = exports.DEFAULT_ITEM_ANIMATION_DURATION = exports.DEFAULT_ITEM_ANIMATION_EASING = void 0;

var _utilities = require("../../utilities");

const DEFAULT_ITEM_ANIMATION_EASING = _utilities.Easing.out(_utilities.Easing.exp);

exports.DEFAULT_ITEM_ANIMATION_EASING = DEFAULT_ITEM_ANIMATION_EASING;
const DEFAULT_ITEM_ANIMATION_DURATION = 500;
exports.DEFAULT_ITEM_ANIMATION_DURATION = DEFAULT_ITEM_ANIMATION_DURATION;
const DEFAULT_ITEM_INNER_SPACE = 12;
exports.DEFAULT_ITEM_INNER_SPACE = DEFAULT_ITEM_INNER_SPACE;
const DEFAULT_ITEM_OUTER_SPACE = 12;
exports.DEFAULT_ITEM_OUTER_SPACE = DEFAULT_ITEM_OUTER_SPACE;
const DEFAULT_ITEM_ICON_SIZE = 24;
exports.DEFAULT_ITEM_ICON_SIZE = DEFAULT_ITEM_ICON_SIZE;
const DEFAULT_ITEM_LAYOUT_DIRECTION = false;
exports.DEFAULT_ITEM_LAYOUT_DIRECTION = DEFAULT_ITEM_LAYOUT_DIRECTION;
const DEFAULT_ITEM_CONTAINER_WIDTH = 'auto';
exports.DEFAULT_ITEM_CONTAINER_WIDTH = DEFAULT_ITEM_CONTAINER_WIDTH;
