// @gorhom/animated-tabbar/lib/module/AnimatedTabBarView.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { useMemo, useEffect, useRef } from 'react';
import Animated from 'react-native-reanimated';
import { useValue } from 'react-native-redash/lib/module/v1';
import Presets from './presets';
const {
  proc,
  call
} = Animated;
/**
 * @DEV
 * this is needed for react-native-svg to animate on the native thread.
 * @external (https://github.com/software-mansion/react-native-reanimated/issues/537)
 */

Animated.addWhitelistedNativeProps({
  width: true,
  stroke: true,
  backgroundColor: true
});
export function AnimatedTabBarView(props) {
  // props
  const {
    index: controlledIndex,
    onIndexChange,
    onLongPress,
    tabs: _tabs,
    preset = 'bubble',
    style,
    itemInnerSpace,
    itemOuterSpace,
    itemContainerWidth,
    iconSize,
    duration,
    easing,
    isRTL
  } = props,
        rest = _objectWithoutProperties(props, ["index", "onIndexChange", "onLongPress", "tabs", "preset", "style", "itemInnerSpace", "itemOuterSpace", "itemContainerWidth", "iconSize", "duration", "easing", "isRTL"]); // verify props


  if (!Object.keys(Presets).includes(preset)) {
    throw new Error("Wrong preset been provided. expected one of these: [".concat(Object.keys(Presets).join(', '), "], but found \"").concat(preset, "\"."));
  } // variables


  const selectedIndex = useValue(controlledIndex);
  const tabs = useMemo(() => {
    return Object.keys(_tabs).map(key => {
      return _tabs[key].title && _tabs[key].key ? _tabs[key] : _objectSpread({
        title: key,
        key: "tab-".concat(key)
      }, _tabs[key]);
    });
  }, [_tabs]); //#region Effects

  const indexRef = useRef(controlledIndex);
  /**
   * @DEV
   * here we listen to the controlled index and update
   * selectedIndex value.
   */

  useEffect(() => {
    selectedIndex.setValue(controlledIndex); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [controlledIndex]);
  /**
   * @DEV
   * here we listen to selectedIndex and call `onIndexChange`
   */

  const animatedOnChange = useMemo(() => proc(index => call([index], args => {
    if (onIndexChange) {
      indexRef.current = args[0];
      onIndexChange(args[0]);
    }
  })), // eslint-disable-next-line react-hooks/exhaustive-deps
  []); //#endregion

  const PresetComponent = Presets[preset].component; // render

  return /*#__PURE__*/React.createElement(PresetComponent, _extends({
    style: style,
    selectedIndex: selectedIndex,
    animatedOnChange: animatedOnChange // @ts-ignore
    ,
    tabs: tabs,
    itemInnerSpace: itemInnerSpace,
    itemOuterSpace: itemOuterSpace,
    itemContainerWidth: itemContainerWidth,
    iconSize: iconSize,
    duration: duration,
    easing: easing,
    isRTL: isRTL,
    onLongPress: onLongPress
  }, rest));
}
