// @gorhom/animated-tabbar/lib/module/presets/flashy/FlashyTabBar.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { useMemo, memo } from 'react';
import { View } from 'react-native'; // @ts-ignore 😞

import isEqual from 'lodash.isequal';
import FlashyTabBarItem from './item';
import RawButton from '../../components/rawButton';
import { DEFAULT_ITEM_ANIMATION_DURATION, DEFAULT_ITEM_ANIMATION_EASING, DEFAULT_ITEM_INNER_SPACE, DEFAULT_ITEM_OUTER_SPACE, DEFAULT_ITEM_ICON_SIZE, DEFAULT_ITEM_LAYOUT_DIRECTION, DEFAULT_ITEM_CONTAINER_WIDTH } from './constants';
import { useTabBarItemFocusTransition, useTabBarItemSpacing } from '../../hooks';
import { noop } from '../../utilities';
import { styles } from './styles';

const FlashyTabBarComponent = ({
  selectedIndex,
  tabs,
  duration = DEFAULT_ITEM_ANIMATION_DURATION,
  easing = DEFAULT_ITEM_ANIMATION_EASING,
  itemInnerSpace,
  itemOuterSpace,
  itemContainerWidth = DEFAULT_ITEM_CONTAINER_WIDTH,
  iconSize = DEFAULT_ITEM_ICON_SIZE,
  isRTL = DEFAULT_ITEM_LAYOUT_DIRECTION,
  style: containerStyleOverride,
  onLongPress = noop,
  animatedOnChange
}) => {
  //#region variables
  const animatedFocusValues = useMemo(() => tabs.map((_, index) => // eslint-disable-next-line react-hooks/rules-of-hooks
  useTabBarItemFocusTransition({
    index,
    selectedIndex,
    duration,
    easing
  })), // eslint-disable-next-line react-hooks/exhaustive-deps
  [tabs, duration, easing]);
  const tabBarItemSpacing = useTabBarItemSpacing(itemInnerSpace, itemOuterSpace, DEFAULT_ITEM_INNER_SPACE, DEFAULT_ITEM_OUTER_SPACE); //#endregion
  //#region styles

  const containerStyle = useMemo(() => [styles.container, containerStyleOverride, {
    flexDirection: isRTL ? 'row-reverse' : 'row'
  }], [containerStyleOverride, isRTL]);
  const rawButtonStyle = useMemo(() => itemContainerWidth === 'fill' ? {
    flex: 1
  } : {}, [itemContainerWidth]); //#endregion
  // render

  return /*#__PURE__*/React.createElement(View, {
    style: containerStyle
  }, tabs.map((_ref, index) => {
    let {
      key,
      title
    } = _ref,
        configs = _objectWithoutProperties(_ref, ["key", "title"]);

    return /*#__PURE__*/React.createElement(RawButton, {
      key: key,
      index: index,
      selectedIndex: selectedIndex,
      style: rawButtonStyle,
      accessibilityLabel: title,
      animatedOnChange: animatedOnChange,
      onLongPress: onLongPress
    }, /*#__PURE__*/React.createElement(FlashyTabBarItem, _extends({
      index: index,
      animatedFocus: animatedFocusValues[index],
      label: title,
      spacing: tabBarItemSpacing,
      itemContainerWidth: itemContainerWidth,
      iconSize: iconSize,
      isRTL: isRTL
    }, configs)));
  }));
};

const FlashyTabBar = /*#__PURE__*/memo(FlashyTabBarComponent, (prevProps, nextProps) => isEqual(prevProps, nextProps));
export default FlashyTabBar;
