// @gorhom/animated-tabbar/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _AnimatedTabBar.AnimatedTabBar;
  }
});
Object.defineProperty(exports, "AnimatedTabBarView", {
  enumerable: true,
  get: function get() {
    return _AnimatedTabBarView.AnimatedTabBarView;
  }
});
Object.defineProperty(exports, "TabsConfig", {
  enumerable: true,
  get: function get() {
    return _types.TabsConfig;
  }
});
Object.defineProperty(exports, "BubbleTabBarConfig", {
  enumerable: true,
  get: function get() {
    return _bubble.BubbleTabBarConfig;
  }
});
Object.defineProperty(exports, "BubbleTabBarItemConfig", {
  enumerable: true,
  get: function get() {
    return _bubble.BubbleTabBarItemConfig;
  }
});
Object.defineProperty(exports, "BubbleTabBarItem", {
  enumerable: true,
  get: function get() {
    return _bubble.BubbleTabBarItem;
  }
});
Object.defineProperty(exports, "BubbleTabBarItemProps", {
  enumerable: true,
  get: function get() {
    return _bubble.BubbleTabBarItemProps;
  }
});
Object.defineProperty(exports, "BubbleTabBarIconProps", {
  enumerable: true,
  get: function get() {
    return _bubble.BubbleTabBarIconProps;
  }
});
Object.defineProperty(exports, "FlashyTabBarConfig", {
  enumerable: true,
  get: function get() {
    return _flashy.FlashyTabBarConfig;
  }
});
Object.defineProperty(exports, "FlashyTabBarItemConfig", {
  enumerable: true,
  get: function get() {
    return _flashy.FlashyTabBarItemConfig;
  }
});
Object.defineProperty(exports, "FlashyTabBarItem", {
  enumerable: true,
  get: function get() {
    return _flashy.FlashyTabBarItem;
  }
});
Object.defineProperty(exports, "FlashyTabBarItemProps", {
  enumerable: true,
  get: function get() {
    return _flashy.FlashyTabBarItemProps;
  }
});
Object.defineProperty(exports, "FlashyTabBarIconProps", {
  enumerable: true,
  get: function get() {
    return _flashy.FlashyTabBarIconProps;
  }
});
Object.defineProperty(exports, "MaterialTabBarConfig", {
  enumerable: true,
  get: function get() {
    return _material.MaterialTabBarConfig;
  }
});
Object.defineProperty(exports, "MaterialTabBarItemConfig", {
  enumerable: true,
  get: function get() {
    return _material.MaterialTabBarItemConfig;
  }
});
Object.defineProperty(exports, "MaterialTabBarItem", {
  enumerable: true,
  get: function get() {
    return _material.MaterialTabBarItem;
  }
});
Object.defineProperty(exports, "MaterialTabBarItemProps", {
  enumerable: true,
  get: function get() {
    return _material.MaterialTabBarItemProps;
  }
});
Object.defineProperty(exports, "MaterialTabBarIconProps", {
  enumerable: true,
  get: function get() {
    return _material.MaterialTabBarIconProps;
  }
});

var _AnimatedTabBar = require("./AnimatedTabBar");

var _AnimatedTabBarView = require("./AnimatedTabBarView");

var _types = require("./types");

var _bubble = require("./presets/bubble");

var _flashy = require("./presets/flashy");

var _material = require("./presets/material");
